pkg/util/etcdutil/migrate.go

```diff
@@ -136,41 +138,36 @@ func createMigratedEtcdCluster(restclient restclient.Interface, host, podIP stri
 	return nil
 }
 
-func checkEtcdClusterUp() error {
+func waitEtcdClusterRunning(restclient restclient.Interface, host string) error {
 	glog.Infof("initial delaying (30s)...")
 	time.Sleep(30 * time.Second)
 
-	// The checking does:
-	// - Trying to talk to etcd cluster via etcd service. The assumption here is that
-	//   the etcd service only selects the etcd pods newly created, excluding the boot one.
-	//   Once we can talk to it, we are sure that etcd cluster is created.
-	// - Then we list members and see if it's been reduced to 1 member. Because when we
-	//   can talk to the etcd cluster, we are certain there are 2 members at the beginning,
-	//   and will reduce to 1 eventually. That's the timeline of expected events.
-	//   As long as 1 member cluster is reached, we are certain cluster has been migrated successfully.
-	err := wait.PollImmediate(10*time.Second, 60*time.Second, func() (bool, error) {
-		cfg := clientv3.Config{
-			Endpoints:   []string{fmt.Sprintf("http://%s:2379", etcdServiceIP)},
-			DialTimeout: 5 * time.Second,
+	err := wait.Poll(10*time.Second, 300*time.Second, func() (bool, error) {
+		uri := fmt.Sprintf("%s/apis/coreos.com/v1/namespaces/kube-system/etcdclusters/%s", host, etcdClusterName)
+		res := restclient.Get().RequestURI(uri).Do()
+		if res.Error() != nil {
+			return false, res.Error()
 		}
-		etcdcli, err := clientv3.New(cfg)
-		if err != nil {
-			glog.Errorf("fail to create etcd client, retrying...: %v", err)
-			return false, nil
+
+		var status int
+		res.StatusCode(&status)
+
+		if status != http.StatusOK {
+			return false, fmt.Errorf("invalid status code: %v", status)
 		}
-		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
-		defer cancel()
-		m, err := etcdcli.MemberList(ctx)
+
+		e := &spec.EtcdCluster{}
+		err := res.Into(e)
 		if err != nil {
-			glog.Errorf("fail to list etcd members, retrying...: %v", err)
-			return false, nil
+			return false, err
 		}
-		if len(m.Members) != 1 {
-			glog.Infof("Still migrating boot etcd member, retrying...")
-			return false, nil
+		switch e.Status.Phase {
+		case spec.ClusterPhaseRunning:
+			return true, nil
+		case spec.ClusterPhaseFailed:
+			return false, errors.New("failed to create etcd cluster")
```

Comment:
You can also use fmt.Errorf() if you want to be consistent, but you can leave as is.